pkg/state/state_macvlan_network.go

```diff
@@ -0,0 +1,217 @@
+)
+
+// NewStateMacvlanNetwork creates a new state for MacvlanNetwork CR
+func NewStateMacvlanNetwork(k8sAPIClient client.Client, scheme *runtime.Scheme, manifestDir string) (State, error) {
+	files, err := utils.GetFilesWithSuffix(manifestDir, render.ManifestFileSuffix...)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get files from manifest dir")
+	}
+
+	renderer := render.NewRenderer(files)
+	return &stateMacvlanNetwork{
+		stateSkel: stateSkel{
+			name:        stateMacvlanNetworkName,
+			description: stateMacvlanNetworkDescription,
+			client:      k8sAPIClient,
+			scheme:      scheme,
+			renderer:    renderer,
+		}}, nil
+}
+
+type stateMacvlanNetwork struct {
+	stateSkel
+}
+
+// Sync attempt to get the system to match the desired state which State represent.
+// a sync operation must be relatively short and must not block the execution thread.
+func (s *stateMacvlanNetwork) Sync(customResource interface{}, _ InfoCatalog) (SyncState, error) {
+	cr := customResource.(*mellanoxv1alpha1.MacvlanNetwork)
+	log.V(consts.LogLevelInfo).Info(
+		"Sync Custom resource", "State:", s.name, "Name:", cr.Name, "Namespace:", cr.Namespace)
+
+	objs, err := s.getManifestObjects(cr)
+	if err != nil {
+		cr.Status.Reason = "failed to render MacvlanNetwork"
+		return SyncStateError, errors.Wrap(err, "failed to render MacvlanNetwork")
+	}
+
+	if len(objs) == 0 {
+		cr.Status.Reason = "no rendered objects found"
+		return SyncStateError, errors.Wrap(err, "no rendered objects found")
+	}
+
+	netAttDef := objs[0]
+	if netAttDef.GetKind() != "NetworkAttachmentDefinition" {
+		cr.Status.Reason = "no NetworkAttachmentDefinition object found"
+		return SyncStateError, errors.Wrap(err, "no NetworkAttachmentDefinition object found")
+	}
+
+	// Delete NetworkAttachmentDefinition if not in desired namespace
+	if err = s.handleNamespaceChange(cr, netAttDef); err != nil {
+		cr.Status.Reason = "Couldn't delete NetworkAttachmentDefinition CR"
+		return SyncStateError, errors.Wrap(err, "Couldn't delete NetworkAttachmentDefinition CR")
+	}
+
+	// TODO modify createAndUpdateObjs() logic to set resource version for objects
+
+	// Set MacvlanNetwork CR the ControllerReference NetworkAttachmentDefinition CR
+	if err := controllerutil.SetControllerReference(cr, netAttDef, s.scheme); err != nil {
+		cr.Status.Reason = "failed to set controller reference for object"
+		return SyncStateError, errors.Wrap(err, "failed to set controller reference for object")
+	}
+
+	// Create objects if they don't exist or moved to a new namespace or update if exists
+	if err := s.createOrUpdate(cr, netAttDef); err != nil {
+		return SyncStateError, err
+	}
+
+	// Check objects status
+	syncState, err := s.getSyncState(objs)
+	if err != nil {
+		return SyncStateNotReady, errors.Wrap(err, "failed to get sync state")
+	}
+
+	// Get NetworkAttachmentDefinition SelfLink
+	if err := s.getObj(netAttDef); err != nil {
+		cr.Status.Reason = "failed to get NetworkAttachmentDefinition"
```

Comment:
I missed that with all the changes going on in this PR

im not too happy with handling the global CR Status in the State object.
There may be multiple States for a controller.

As there is only one State its OK for now.

Reply:
can you add a TODO here saying that CR `Status` handling should be done in the controller ?

Reply:
mind open an issue on this so we can track?